drm dumb allocator

bl4ckb0ne opened the pull request.

emersion commented on include/render/drm_dumb_allocator.h:
```diff
@@ -0,0 +1,32 @@
+	int drm_fd;
+
+	uint32_t id;
+	uint32_t handles[4];
```
Dumb buffers only ever have a single plane, so we don't need arrays here.

bl4ckb0ne replied:
Is it the same for the strides and offsets?

emersion replied:
Yeah

bl4ckb0ne replied:
I removed `offsets` since it's always 0.

emersion commented on render/drm_dumb_allocator.c:
```diff
@@ -0,0 +1,183 @@
+	struct drm_mode_create_dumb create = {0};
+	create.width = (uint32_t)width;
+	create.height = (uint32_t)height;
+	create.bpp = 32;
```
We should probably only support ARGB8888 and XRGB8888 here.

bl4ckb0ne replied:
So this should be a constant somewhere and the format should be checked right? Would it be possible to have a different bpp value with other formats?

emersion replied:
Yeah, there's a direct mapping between DRM formats and bpp for dumb buffers. 32 means ARGB8888 and 24 means XRGB8888. A simple switch is probably enough.

bl4ckb0ne replied:
Should I make it futureproof? Or we just wont support any other format

bl4ckb0ne replied:
As discussed on IRC, ARGB8888 and XRGB8888 both have a BPP of 32, if the format requested is not one of those, the creation of the buffer should fail gracefully.

bl4ckb0ne replied:
Done.

emersion commented on render/drm_dumb_allocator.c:
```diff
@@ -0,0 +1,213 @@
+		uint32_t *size) {
+	struct wlr_drm_dumb_buffer *buf = drm_dumb_buffer_from_buffer(wlr_buffer);
+	*data = buf->data;
+	memcpy(size, &buf->size, sizeof(buf->size));
```
This can just be `*size = buf->size`.

emersion commented on render/drm_dumb_allocator.c:
```diff
@@ -0,0 +1,213 @@
+}
+
+static void finish_buffer(struct wlr_drm_dumb_buffer *buf) {
+
```
Nit: stray newline

emersion commented on include/wlr/types/wlr_buffer.h:
```diff
@@ -76,6 +78,14 @@ void wlr_buffer_unlock(struct wlr_buffer *buffer);
+ * The returned pointer should be pointing to a valid memory location.
+ */
+bool wlr_buffer_get_data_ptr(struct wlr_buffer *buffer, void **data,
+	uint32_t *size);
```
`size` should probably be a `size_t *`, it's the native C type for… sizes.

emersion commented on types/wlr_buffer.c:
```diff
@@ -64,6 +64,13 @@ bool wlr_buffer_get_dmabuf(struct wlr_buffer *buffer,
 }
 
+bool wlr_buffer_get_data_ptr(struct wlr_buffer *buffer, void **data,
+	uint32_t *size) {
```
Style: need one extra indentation level for this continuation line, otherwise we can't tell it apart from the function body.

emersion commented on include/wlr/types/wlr_buffer.h:
```diff
@@ -76,6 +78,14 @@ void wlr_buffer_unlock(struct wlr_buffer *buffer);
+ *
+ * The returned pointer should be pointing to a valid memory location.
+ */
+bool wlr_buffer_get_data_ptr(struct wlr_buffer *buffer, void **data,
```
Can we move this into an internal header for now? (e.g. `include/types/wlr_buffer.h`)

I'm not yet 100% sure this is the right interface, not making it public right away allows us to more easily change it if necessary.

bl4ckb0ne replied:
Sure

emersion commented on include/types/wlr_buffer.h:
```diff
@@ -0,0 +1,9 @@
+#include "wlr/types/wlr_buffer.h"
```
This file needs include guards (`#ifndef TYPES_WLR_BUFFER_H`).

bl4ckb0ne replied:
Oh I took example from `types/wlr_keyboard.h` and it seems to be the only one that doesnt have an include guard.

emersion replied:
Oh, oops, that's an oversight. Can you send a patch to fix `wlr_keyboard.h`?

bl4ckb0ne replied:
Sure!

emersion commented on include/types/wlr_buffer.h:
```diff
@@ -0,0 +1,9 @@
+ * Access a pointer to the allocated data from the underlying implementation,
+ * and its stride.
+ *
+ * The returned pointer should be pointing to a valid memory location.
```
Can we state that the pointer is valid for both read and write operations?

emersion commented on render/drm_dumb_allocator.c:
```diff
@@ -0,0 +1,212 @@
+#define _POSIX_C_SOURCE 200809L
+#include "render/drm_dumb_allocator.h"
```
Style: the order of these includes isn't consistent with the rest of wlroots. In general we do first the angle-bracket includes (sorted alphabetically) then the double-quote includes (sorted alphabetically).

emersion commented on render/drm_dumb_allocator.c:
```diff
@@ -0,0 +1,212 @@
+	int prime_fd;
+	if (drmPrimeHandleToFD(alloc->drm_fd, buffer->handle, DRM_CLOEXEC,
+			&prime_fd) != 0) {
+		wlr_log(WLR_ERROR, "failed to get prime handle from GEM: %s",
```
Can use `wlr_log_errno` here.

emersion commented on render/drm_dumb_allocator.c:
```diff
@@ -0,0 +1,212 @@
+		goto create_destroy;
+	}
+
+	buffer->data = mmap(NULL, create.size, PROT_WRITE, MAP_SHARED,
```
Ah, this is a write-only mapping. Need to document that.

emersion commented on include/render/drm_dumb_allocator.h:
```diff
@@ -0,0 +1,32 @@
+	int drm_fd;
+};
+
+struct wlr_drm_dumb_allocator *wlr_drm_dumb_allocator_create(int drm_fd);
```
Can we document this function a bit? I think we should make it clear that:

- It takes ownership of the DRM FD
- When importing DMA-BUFs coming from this allocator, the caller must not use the same DRM file description, because GEM handles are not reference-counted.

emersion commented on render/drm_dumb_allocator.c:
```diff
@@ -0,0 +1,212 @@
+
+	if (buf->drm_fd >= -1) {
+		struct drm_mode_destroy_dumb destroy = { .handle = buf->handle };
+		drmIoctl(buf->drm_fd, DRM_IOCTL_MODE_DESTROY_DUMB, &destroy);
```
It would be nice to print a log on error.

emersion commented on render/drm_dumb_allocator.c:
```diff
@@ -0,0 +1,212 @@
+static struct wlr_buffer *allocator_create_buffer(
+		struct wlr_allocator *wlr_alloc, int width, int height,
+		const struct wlr_drm_format *drm_format) {
+	if (drm_format->format != DRM_FORMAT_ARGB8888
```
Maybe we can move this check in `create_buffer`? It would be closed to the assumption `bpp = 32`.

bl4ckb0ne replied:
Yup, this will have to change one #2779 is merged.

emersion commented on include/types/wlr_buffer.h:
```diff
@@ -0,0 +1,15 @@
+#ifndef TYPES_WLR_BUFFER
+#define TYPES_WLR_BUFFER
+
+#include "wlr/types/wlr_buffer.h"
```
This is a public header, should be included with angle-brackets.

emersion commented:
I think "buffer: introduce wlr_buffer_get_data_ptr" and "buffer: move wlr_buffer_get_data to internal header" can be squashed together.

emersion commented on render/drm_dumb_allocator.c:
```diff
@@ -0,0 +1,201 @@
+	const struct wlr_pixel_format_info *info = drm_get_pixel_format_info(
+			format->format);
+	if (info == NULL) {
+		wlr_log(WLR_ERROR, "drm format not supported");
```
We need to return NULL here

emersion commented on render/drm_dumb_allocator.c:
```diff
@@ -0,0 +1,202 @@
+		munmap(buf->data, buf->size);
+	}
+
+	if (buf->drm_fd >= -1) {
```
This should be `>= 0`. -1 is not a valid FD.

emersion commented:
This looks good to me, but isn't used yet in wlroots. I'll have a look at the pixman renderer, just to make sure we haven't overlooked anything.

emersion commented:
Needs a rebase on top of master.

emersion commented:
I think for this to be merged we just need to:

- Add a way to force Pixman renderer in `wlr_renderer_autocreate` (`WLR_RENDERER` env)
- Check if the renderer is Pixman [here](https://github.com/swaywm/wlroots/blob/78b94a570c7b3050bb3607d82c7e770419be6411/backend/drm/renderer.c#L44) and use the DRM dumb alloc in this case.

emersion commented:
This PR is now blocked on https://github.com/swaywm/wlroots/pull/2884.

bl4ckb0ne commented:
Rebased and updated, I was able to run sway with `WLR_RENDERER=pixman` without any issue.

Small thing I noticed, I'm not able to run the examples using the gles2 renderer from the pixman renderer, is that something we'd like to fix?

emersion commented on render/allocator.c:
```diff
@@ -23,21 +24,35 @@ struct wlr_allocator *wlr_allocator_autocreate(struct wlr_backend *backend,
 	int drm_fd = wlr_backend_get_drm_fd(backend);
+	int fd = -1;
+	if (drm_fd != -1) {
+		fd = fcntl(drm_fd, F_DUPFD_CLOEXEC, 0);
```
Because the DRM dumb allocator is using `drmPrimeHandleToFD`, we can't `dup` the DRM FD here, or we risk double-closing GEM handles. See https://gitlab.freedesktop.org/mesa/drm/-/merge_requests/110 for more info.

Instead, we need to re-open the DRM FD when using the DRM dumb allocator. To do so, we can use `drmGetDeviceNameFromFd2` to get the path to the DRM node, use `wlr_backend_get_session` to get the `wlr_session` (fail if there's no session), and then use `wlr_session_open_file`. Maybe all of this mess can be isolated in a `attempt_drm_dumb_allocator` function or something.

bl4ckb0ne replied:
I took that bit of code from all of the backend, they previously did that operation before creating their allocators. What was preventing double closing GEM handles?

emersion replied:
Other backends and allocators don't use `drmPrimeHandleToFD`, so they can just `dup` the DRM FD. Since the DRM dumb allocator uses `drmPrimeHandleToFD`, it needs to re-open the DRM FD.

emersion commented on render/allocator.c:
```diff
@@ -23,21 +24,35 @@ struct wlr_allocator *wlr_allocator_autocreate(struct wlr_backend *backend,
 
+	uint32_t drm_caps = WLR_BUFFER_CAP_DMABUF | WLR_BUFFER_CAP_DATA_PTR;
+	if ((backend_caps & drm_caps) && (renderer_caps & drm_caps)
+			&& fd != -1) {
```
This will likely result in the DRM dumb allocator being picked when running the Pixman renderer with the Wayland or X11 backend. We should probably try the shm allocator first.

emersion commented:
> I'm not able to run the examples using the gles2 renderer from the pixman renderer, is that something we can fix?

This is a known issue. When running inside a compositor that uses Pixman, Wayland clients won't have access to the GPU, they will fallback to software rendering. llvmpipe is not supported in wlroots in this case, see https://github.com/swaywm/wlroots/pull/2891.

emersion commented:
@J0nnyMak0 said this doesn't work on logind: https://github.com/swaywm/wlroots/issues/2910#issuecomment-831316498

J0nnyMak0 commented:
@bl4ckb0ne now I'm able to start with pixman, however external monitors remain off:
```
00:00:00.004 [DEBUG] [sway/server.c:47] Preparing Wayland server initialization
00:00:00.004 [ERROR] [wlr] [libseat] [libseat/backend/seatd.c:78] Could not connect to socket /run/seatd.sock: No such file or directory
00:00:00.004 [INFO] [wlr] [libseat] [libseat/libseat.c:70] Backend 'seatd' failed to open seat, skipping
00:00:00.010 [INFO] [wlr] [libseat] [libseat/libseat.c:67] Seat opened with backend 'logind'
00:00:00.010 [INFO] [wlr] [backend/session/session.c:110] Successfully loaded libseat session
00:00:00.759 [INFO] [wlr] [backend/backend.c:175] Found 2 GPUs
00:00:00.759 [INFO] [wlr] [backend/drm/backend.c:182] Initializing DRM backend for /dev/dri/card0 (i915)
00:00:00.759 [DEBUG] [wlr] [backend/drm/drm.c:68] Using atomic DRM interface
00:00:00.759 [DEBUG] [wlr] [backend/drm/drm.c:81] ADDFB2 modifiers supported
00:00:00.759 [INFO] [wlr] [backend/drm/drm.c:243] Found 3 DRM CRTCs 
00:00:00.760 [INFO] [wlr] [backend/drm/drm.c:170] Found 9 DRM planes
00:00:00.769 [INFO] [wlr] [render/wlr_renderer.c:237] Loading user-specified renderer due to WLR_RENDERER: pixman
00:00:00.769 [INFO] [wlr] [render/pixman/renderer.c:433] Creating pixman renderer
00:00:00.769 [DEBUG] [wlr] [render/allocator.c:49] Trying to create drm dumb allocator
00:00:00.769 [DEBUG] [wlr] [render/drm_dumb_allocator.c:224] Created drm dumb allocator with DRM node /dev/dri/card0
00:00:00.769 [INFO] [wlr] [backend/drm/backend.c:182] Initializing DRM backend for /dev/dri/card1 (nouveau)
00:00:00.769 [DEBUG] [wlr] [backend/drm/drm.c:64] Atomic modesetting unsupported, using legacy DRM interface
00:00:00.769 [DEBUG] [wlr] [backend/drm/drm.c:81] ADDFB2 modifiers supported
00:00:00.769 [INFO] [wlr] [backend/drm/drm.c:243] Found 4 DRM CRTCs
00:00:01.102 [INFO] [wlr] [backend/drm/drm.c:170] Found 12 DRM planes
00:00:01.141 [INFO] [wlr] [render/wlr_renderer.c:237] Loading user-specified renderer due to WLR_RENDERER: pixman
00:00:01.141 [INFO] [wlr] [render/pixman/renderer.c:433] Creating pixman renderer
00:00:01.141 [DEBUG] [wlr] [render/allocator.c:49] Trying to create drm dumb allocator
00:00:01.142 [DEBUG] [wlr] [render/drm_dumb_allocator.c:224] Created drm dumb allocator with DRM node /dev/dri/card1
00:00:01.142 [INFO] [sway/main.c:358] Starting sway version 1.6-f8d7c170 (May  3 2021, branch 'master')
00:00:01.142 [DEBUG] [sway/server.c:60] Initializing Wayland server
00:00:01.142 [DEBUG] [wlr] [types/wlr_idle.c:247] idle manager created
00:00:01.142 [INFO] [wlr] [backend/noop/backend.c:59] Creating noop backend
00:00:01.142 [INFO] [wlr] [backend/headless/backend.c:241] Creating headless backend with parent renderer
00:00:01.142 [ERROR] [wlr] [backend/headless/backend.c:253] Failed to get DRM device FD from renderer
00:00:01.142 [DEBUG] [wlr] [render/allocator.c:39] Trying to create shm allocator 
00:00:01.142 [DEBUG] [wlr] [render/shm_allocator.c:108] Created shm allocator
```

emersion commented:
Ah, this is a multi-GPU setup. I think it _should_ work thanks to the `read_pixels` Pixman impl, but will likely be pretty slow (can be improved in the future). The logs you posted don't contain anything interesting, but seem truncated, can you post full logs?

emersion commented on include/render/drm_dumb_allocator.h:
```diff
@@ -0,0 +1,41 @@
+ * the same DRM file description, because GEM handles are not
+ * reference-counted.
+ *
+ * Takes ownership over the FD.
```
This doc comment needs to be updated.

emersion commented on render/allocator.c:
```diff
@@ -29,10 +28,7 @@ struct wlr_allocator *allocator_autocreate_with_drm_fd(
-		if (fd < 0) {
-			wlr_log(WLR_ERROR, "fcntl(F_DUPFD_CLOEXEC) failed");
-		} else if ((alloc = wlr_gbm_allocator_create(fd)) != NULL) {
+		if ((alloc = wlr_gbm_allocator_create(drm_fd)) != NULL) {
```
This breaks the GBM allocator temporarily, it's fixed in the next commit. This will make git-bisect operations more difficult, instead we should move this hunk to the next commit.

emersion commented on render/drm_dumb_allocator.c:
```diff
@@ -0,0 +1,228 @@
+		.n_planes = 1,
+		.offset[0] = 0,
+		.stride[0] = buffer->stride,
+		.fd[0] = prime_fd,
```
The PRIME FD is never closed.

bl4ckb0ne replied:
Shouldn't `wlr_dmabuf_attributes_finish` close the fd?

emersion replied:
Yeah I added it

emersion commented on render/drm_dumb_allocator.c:
```diff
@@ -0,0 +1,228 @@
+};
+
+struct wlr_allocator *wlr_drm_dumb_allocator_create(int fd) {
+	char *path = drmGetDeviceNameFromFd2(fd);
```
Would be nice to add a little comment to explain why we do this.

emersion commented on render/gbm_allocator.c:
```diff
@@ -159,7 +159,13 @@ static struct wlr_gbm_allocator *get_gbm_alloc_from_alloc(
 
-struct wlr_allocator *wlr_gbm_allocator_create(int fd) {
+struct wlr_allocator *wlr_gbm_allocator_create(int drm_fd) {
+	int fd = fcntl(drm_fd, F_DUPFD_CLOEXEC, 0);
```
We need to update the function doc-comment, which still states that the function takes ownership over the FD.

emersion commented:
I';ve pushed an updated version that addresses these comments. The previous version is saved in this branch: https://github.com/emersion/wlroots/tree/drm_dumb_allocator_old

emersion commented:
Hm. This doesn't seem to work on my setup:

```
00:00:01.178 [DEBUG] [wlr] [render/drm_dumb_allocator.c:116] Allocated 1680 x 1050 DRM dumb buffer (GEM 1)
00:00:01.178 [DEBUG] [wlr] [render/pixman/renderer.c:113] Created pixman buffer 1680x1050
00:00:01.197 [DEBUG] [wlr] [backend/drm/renderer.c:405] Failed to import DMA-BUF in GBM
00:00:01.197 [ERROR] [wlr] [backend/drm/renderer.c:317] Failed to import buffer
00:00:01.197 [ERROR] [wlr] [backend/drm/drm.c:677] connector DP-1: Failed to initialize renderer:initial page-flip failed
00:00:01.197 [ERROR] [wlr] [backend/drm/drm.c:777] connector DP-1: Failed to initialize renderer for plane
00:00:01.197 [ERROR] [sway/config/output.c:417] Failed to commit output DP-1
```

emersion commented:
Opened https://github.com/swaywm/wlroots/issues/2912 for the `wlr_session` issue.

J0nnyMak0 commented:
>can you post full logs?

Attached.
[sway.log](https://github.com/swaywm/wlroots/files/6421342/sway.log)

emersion commented:
OK, your failure looks different from mine. Yours is:

```
00:00:03.293 [DEBUG] [wlr] [render/drm_dumb_allocator.c:114] Allocated 3840 x 2160 DRM dumb buffer (GEM 14)
00:00:03.293 [DEBUG] [wlr] [render/pixman/renderer.c:113] Created pixman buffer 3840x2160
00:00:03.294 [ERROR] [wlr] [backend/drm/renderer.c:307] Failed to blit buffer across GPUs
00:00:03.294 [ERROR] [wlr] [backend/drm/drm.c:714] connector DP-3: Failed to initialize renderer:initial page-flip failed
00:00:03.294 [ERROR] [wlr] [backend/drm/drm.c:784] connector DP-3: Failed to initialize renderer for plane
```

It seems like DRM only supports blitting DMA-BUFs for multi-GPU. In `drm_surface_blit`, the backend directly calls `wlr_texture_from_dmabuf`, which doesn't work with Pixman.

So, multi-GPU won't work as-is with Pixman, we'll need a follow-up PR to add support for this.

emersion commented:
Seems like I'm hitting this check on amdgpu: https://gitlab.freedesktop.org/mesa/mesa/-/blob/922f71b819b06cd4acbf14dbb1d1b161bd6e3953/src/gallium/drivers/radeonsi/si_texture.c#L1587

Fixing this is going to be non-trivial. The root cause is that DRM dumb buffers can't be used for rendering on amdgpu because of alignment requirements (dumb buffer is unaligned, radeonsi requires `alignment_log2 == 8`).

I think we're going to need to completely side-step GBM here… Either by passing GEM handles around and hoping for the best, either by managing GEM handles ref'counting ourselves.

emersion commented on render/drm_dumb_allocator.c:
```diff
@@ -0,0 +1,235 @@
+	buffer->dmabuf = (struct wlr_dmabuf_attributes){
+		.width = buffer->width,
+		.height = buffer->height,
+		.format = format->format,
```
We forget to set `.modifier` here. Fixed in my latest push.

emersion reviewed (approved):
Alright, let's merge this, and discuss the amdgpu issue separately. Things aren't as simple as they could be, sadly.

Thanks for your contribution!

emersion commented:
Opened https://github.com/swaywm/wlroots/issues/2916 for the GBM issues.
